{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scratch code to see how many hosts of embl dataset hve been downloaded from uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Proteome_ID</th>\n",
       "      <th>Tax_ID</th>\n",
       "      <th>OSCODE</th>\n",
       "      <th>SUPERREGNUM</th>\n",
       "      <th>#(1)</th>\n",
       "      <th>#(2)</th>\n",
       "      <th>#(3)</th>\n",
       "      <th>Species Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UP000521578</td>\n",
       "      <td>47692</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>14416</td>\n",
       "      <td>0</td>\n",
       "      <td>14442</td>\n",
       "      <td>Menura novaehollandiae (superb lyrebird)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UP000029743</td>\n",
       "      <td>1554474</td>\n",
       "      <td>ROSV2</td>\n",
       "      <td>viruses</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Rosavirus A2 (isolate Human/Gambia/GA7403/2008)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UP000628017</td>\n",
       "      <td>2035344</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>3942</td>\n",
       "      <td>0</td>\n",
       "      <td>3942</td>\n",
       "      <td>Amylibacter cionae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UP001232722</td>\n",
       "      <td>2664220</td>\n",
       "      <td>NaN</td>\n",
       "      <td>viruses</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>Tadarida brasiliensis gemykibivirus 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UP000248311</td>\n",
       "      <td>1795041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>3295</td>\n",
       "      <td>0</td>\n",
       "      <td>3300</td>\n",
       "      <td>Pseudoroseicyclus aestuarii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24950</th>\n",
       "      <td>UP000203542</td>\n",
       "      <td>2560777</td>\n",
       "      <td>NaN</td>\n",
       "      <td>viruses</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>Spheniscid alphaherpesvirus 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24951</th>\n",
       "      <td>UP000273303</td>\n",
       "      <td>2490946</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>2387</td>\n",
       "      <td>0</td>\n",
       "      <td>2391</td>\n",
       "      <td>Actinobaculum sp. 352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24952</th>\n",
       "      <td>UP000031129</td>\n",
       "      <td>743971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>553</td>\n",
       "      <td>0</td>\n",
       "      <td>563</td>\n",
       "      <td>Mesomycoplasma flocculare ATCC 27399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24953</th>\n",
       "      <td>UP000000814</td>\n",
       "      <td>272562</td>\n",
       "      <td>CLOAB</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>3847</td>\n",
       "      <td>0</td>\n",
       "      <td>3847</td>\n",
       "      <td>Clostridium acetobutylicum (strain ATCC 824 / ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24954</th>\n",
       "      <td>UP000000725</td>\n",
       "      <td>155148</td>\n",
       "      <td>BPHK6</td>\n",
       "      <td>viruses</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>Enterobacteria phage HK620 (Bacteriophage HK620)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24955 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Proteome_ID   Tax_ID OSCODE SUPERREGNUM   #(1)  #(2)   #(3)  \\\n",
       "0      UP000521578    47692    NaN   eukaryota  14416     0  14442   \n",
       "1      UP000029743  1554474  ROSV2     viruses      1     0      1   \n",
       "2      UP000628017  2035344    NaN    bacteria   3942     0   3942   \n",
       "3      UP001232722  2664220    NaN     viruses      2     0      2   \n",
       "4      UP000248311  1795041    NaN    bacteria   3295     0   3300   \n",
       "...            ...      ...    ...         ...    ...   ...    ...   \n",
       "24950  UP000203542  2560777    NaN     viruses     83     0     83   \n",
       "24951  UP000273303  2490946    NaN    bacteria   2387     0   2391   \n",
       "24952  UP000031129   743971    NaN    bacteria    553     0    563   \n",
       "24953  UP000000814   272562  CLOAB    bacteria   3847     0   3847   \n",
       "24954  UP000000725   155148  BPHK6     viruses     58     0     58   \n",
       "\n",
       "                                            Species Name  \n",
       "0               Menura novaehollandiae (superb lyrebird)  \n",
       "1        Rosavirus A2 (isolate Human/Gambia/GA7403/2008)  \n",
       "2                                     Amylibacter cionae  \n",
       "3                  Tadarida brasiliensis gemykibivirus 1  \n",
       "4                            Pseudoroseicyclus aestuarii  \n",
       "...                                                  ...  \n",
       "24950                      Spheniscid alphaherpesvirus 1  \n",
       "24951                              Actinobaculum sp. 352  \n",
       "24952               Mesomycoplasma flocculare ATCC 27399  \n",
       "24953  Clostridium acetobutylicum (strain ATCC 824 / ...  \n",
       "24954   Enterobacteria phage HK620 (Bacteriophage HK620)  \n",
       "\n",
       "[24955 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_reference = pd.read_csv(r\"C:\\Users\\lorenzo\\Desktop\\BLM\\bacteria\\uniprot_reference_proteomes_ids.csv\", sep = ';')\n",
    "uniprot_reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phage_ID</th>\n",
       "      <th>Length</th>\n",
       "      <th>GC_content</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Host</th>\n",
       "      <th>Lifestyle</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Subcluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AJ550635.1</td>\n",
       "      <td>4554</td>\n",
       "      <td>41.040843</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>High-quality</td>\n",
       "      <td>Chlamydophila pecorum</td>\n",
       "      <td>virulent</td>\n",
       "      <td>cluster_152500</td>\n",
       "      <td>subcluster_191945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AJ969242.1</td>\n",
       "      <td>7878</td>\n",
       "      <td>23.013455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not-determined</td>\n",
       "      <td>Spiroplasma citri</td>\n",
       "      <td>virulent</td>\n",
       "      <td>cluster_24190</td>\n",
       "      <td>subcluster_30305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FQ482084.1</td>\n",
       "      <td>45522</td>\n",
       "      <td>49.681473</td>\n",
       "      <td>Caudovirales</td>\n",
       "      <td>High-quality</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "      <td>virulent</td>\n",
       "      <td>cluster_223352</td>\n",
       "      <td>subcluster_281280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FR725450.1</td>\n",
       "      <td>23750</td>\n",
       "      <td>52.985263</td>\n",
       "      <td>Caudovirales</td>\n",
       "      <td>Medium-quality</td>\n",
       "      <td>Nitrososphaera viennensis EN76</td>\n",
       "      <td>virulent</td>\n",
       "      <td>cluster_67564</td>\n",
       "      <td>subcluster_85097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AJ536073.2</td>\n",
       "      <td>14931</td>\n",
       "      <td>39.729422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>High-quality</td>\n",
       "      <td>Bacillus thuringiensis serovar israelensis</td>\n",
       "      <td>temperate</td>\n",
       "      <td>cluster_26036</td>\n",
       "      <td>subcluster_32746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>LN681542.1</td>\n",
       "      <td>52261</td>\n",
       "      <td>28.870477</td>\n",
       "      <td>Caudovirales</td>\n",
       "      <td>High-quality</td>\n",
       "      <td>Clostridium difficile</td>\n",
       "      <td>temperate</td>\n",
       "      <td>cluster_66167</td>\n",
       "      <td>subcluster_83369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>HE956708.2</td>\n",
       "      <td>122696</td>\n",
       "      <td>40.512323</td>\n",
       "      <td>Caudovirales</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Yersinia enterocolitica</td>\n",
       "      <td>virulent</td>\n",
       "      <td>cluster_24220</td>\n",
       "      <td>subcluster_30337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>HE956707.2</td>\n",
       "      <td>42085</td>\n",
       "      <td>51.845075</td>\n",
       "      <td>Caudovirales</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Yersinia enterocolitica</td>\n",
       "      <td>virulent</td>\n",
       "      <td>cluster_44487</td>\n",
       "      <td>subcluster_55840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>HE956710.2</td>\n",
       "      <td>42406</td>\n",
       "      <td>47.637127</td>\n",
       "      <td>Caudovirales</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Yersinia enterocolitica</td>\n",
       "      <td>virulent</td>\n",
       "      <td>cluster_44472</td>\n",
       "      <td>subcluster_55821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>LN681537.2</td>\n",
       "      <td>131704</td>\n",
       "      <td>26.415295</td>\n",
       "      <td>Caudovirales</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Clostridium difficile</td>\n",
       "      <td>temperate</td>\n",
       "      <td>cluster_161207</td>\n",
       "      <td>subcluster_202910</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Phage_ID  Length  GC_content      Taxonomy    Completeness  \\\n",
       "0    AJ550635.1    4554   41.040843  Microviridae    High-quality   \n",
       "1    AJ969242.1    7878   23.013455           NaN  Not-determined   \n",
       "2    FQ482084.1   45522   49.681473  Caudovirales    High-quality   \n",
       "3    FR725450.1   23750   52.985263  Caudovirales  Medium-quality   \n",
       "4    AJ536073.2   14931   39.729422           NaN    High-quality   \n",
       "..          ...     ...         ...           ...             ...   \n",
       "151  LN681542.1   52261   28.870477  Caudovirales    High-quality   \n",
       "152  HE956708.2  122696   40.512323  Caudovirales        Complete   \n",
       "153  HE956707.2   42085   51.845075  Caudovirales        Complete   \n",
       "154  HE956710.2   42406   47.637127  Caudovirales        Complete   \n",
       "155  LN681537.2  131704   26.415295  Caudovirales        Complete   \n",
       "\n",
       "                                           Host  Lifestyle         Cluster  \\\n",
       "0                         Chlamydophila pecorum   virulent  cluster_152500   \n",
       "1                             Spiroplasma citri   virulent   cluster_24190   \n",
       "2                             Erwinia amylovora   virulent  cluster_223352   \n",
       "3                Nitrososphaera viennensis EN76   virulent   cluster_67564   \n",
       "4    Bacillus thuringiensis serovar israelensis  temperate   cluster_26036   \n",
       "..                                          ...        ...             ...   \n",
       "151                       Clostridium difficile  temperate   cluster_66167   \n",
       "152                     Yersinia enterocolitica   virulent   cluster_24220   \n",
       "153                     Yersinia enterocolitica   virulent   cluster_44487   \n",
       "154                     Yersinia enterocolitica   virulent   cluster_44472   \n",
       "155                       Clostridium difficile  temperate  cluster_161207   \n",
       "\n",
       "            Subcluster  \n",
       "0    subcluster_191945  \n",
       "1     subcluster_30305  \n",
       "2    subcluster_281280  \n",
       "3     subcluster_85097  \n",
       "4     subcluster_32746  \n",
       "..                 ...  \n",
       "151   subcluster_83369  \n",
       "152   subcluster_30337  \n",
       "153   subcluster_55840  \n",
       "154   subcluster_55821  \n",
       "155  subcluster_202910  \n",
       "\n",
       "[156 rows x 9 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embl_meta = pd.read_csv(r\"C:\\Users\\lorenzo\\Desktop\\datasets\\phagescope\\Metadata\\phage_metadata\\embl_phage_meta_data.tsv\", sep = '\\t')\n",
    "embl_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                           Chlamydophila pecorum\n",
       "1                               Spiroplasma citri\n",
       "2                               Erwinia amylovora\n",
       "3                  Nitrososphaera viennensis EN76\n",
       "4      Bacillus thuringiensis serovar israelensis\n",
       "                          ...                    \n",
       "151                         Clostridium difficile\n",
       "152                       Yersinia enterocolitica\n",
       "153                       Yersinia enterocolitica\n",
       "154                       Yersinia enterocolitica\n",
       "155                         Clostridium difficile\n",
       "Name: Host, Length: 156, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split the \"Host\" column into multiple rows\n",
    "bacteria_series = embl_meta['Host'].str.split(';').explode()\n",
    "\n",
    "# Remove any empty strings that might result from splitting\n",
    "bacteria_series = bacteria_series[bacteria_series != '']\n",
    "bacteria_series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                           Chlamydophila pecorum\n",
       "1                               Spiroplasma citri\n",
       "2                               Erwinia amylovora\n",
       "3                  Nitrososphaera viennensis EN76\n",
       "4      Bacillus thuringiensis serovar israelensis\n",
       "5                          Pseudomonas aeruginosa\n",
       "6                                        Brucella\n",
       "16                          Clostridium difficile\n",
       "17                      Burkholderia pseudomallei\n",
       "20                                   Enterococcus\n",
       "21                               Escherichia coli\n",
       "30                                  Mycobacterium\n",
       "31                          Staphylococcus aureus\n",
       "36                              Bacillus subtilis\n",
       "38                  Lactiplantibacillus plantarum\n",
       "39                       Streptococcus pneumoniae\n",
       "41                          Lactobacillus gasseri\n",
       "42                         Listeria monocytogenes\n",
       "43                        Yersinia enterocolitica\n",
       "45                                       Bacillus\n",
       "46                         Parabacteroides merdae\n",
       "52                                   Streptomyces\n",
       "55                                  Synechococcus\n",
       "56                   Stenotrophomonas maltophilia\n",
       "57                                 Flavobacterium\n",
       "65                                Yersinia pestis\n",
       "69                        Azospirillum brasilense\n",
       "70                            Salmonella enterica\n",
       "71                         Streptococcus gordonii\n",
       "78                           Streptococcus oralis\n",
       "82                        Pseudomonas fluorescens\n",
       "83                        Vibrio parahaemolyticus\n",
       "84                                  Campylobacter\n",
       "86                                Gordonia terrae\n",
       "89                            Pantoea agglomerans\n",
       "91                           Erwinia tasmaniensis\n",
       "93                             Pseudomonas putida\n",
       "95                           Campylobacter jejuni\n",
       "98                           Bacteroides fragilis\n",
       "99                                 Dickeya solani\n",
       "101                       Acinetobacter baumannii\n",
       "104                       Staphylococcus pasteuri\n",
       "105                        Bacillus thuringiensis\n",
       "109      Escherichia coli str. K-12 substr. W3110\n",
       "113                      Lactobacillus helveticus\n",
       "114                    Streptococcus thermophilus\n",
       "120                      Clostridioides difficile\n",
       "122                        Cronobacter turicensis\n",
       "123                         Citrobacter rodentium\n",
       "125                                   Clostridium\n",
       "128                             Vibrio vulnificus\n",
       "Name: Host, dtype: object"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean bacterial names\n",
    "def clean_bacteria_name(name):\n",
    "    name = name.replace('[', '').replace(']', '').strip()  # Remove brackets\n",
    "    if 'sp.' in name:  # Handle species abbreviations\n",
    "        name = name.split('sp.')[0].strip()\n",
    "    return name\n",
    "\n",
    "bacteria_series = bacteria_series.apply(clean_bacteria_name).drop_duplicates()\n",
    "bacteria_series\n",
    "#bacteria_series_cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                 Menura novaehollandiae (superb lyrebird)\n",
       "1          Rosavirus A2 (isolate Human/Gambia/GA7403/2008)\n",
       "2                                       Amylibacter cionae\n",
       "3                    Tadarida brasiliensis gemykibivirus 1\n",
       "4                              Pseudoroseicyclus aestuarii\n",
       "                               ...                        \n",
       "24950                        Spheniscid alphaherpesvirus 1\n",
       "24951                                Actinobaculum sp. 352\n",
       "24952                 Mesomycoplasma flocculare ATCC 27399\n",
       "24953    Clostridium acetobutylicum (strain ATCC 824 / ...\n",
       "24954     Enterobacteria phage HK620 (Bacteriophage HK620)\n",
       "Name: Species Name, Length: 24955, dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_reference['Species Name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32\n",
      "5      UP000002438\n",
      "6      UP000646478\n",
      "17     UP000000605\n",
      "20     UP000014136\n",
      "21     UP000001410\n",
      "30     UP000092151\n",
      "31     UP000008816\n",
      "36     UP000001570\n",
      "38     UP000000432\n",
      "39     UP000000586\n",
      "42     UP000000817\n",
      "45     UP000481043\n",
      "52     UP000198873\n",
      "55     UP000240206\n",
      "56     UP000008840\n",
      "57     UP000557042\n",
      "65     UP000000815\n",
      "70     UP000248731\n",
      "71     UP000001131\n",
      "82     UP000001036\n",
      "84     UP000254920\n",
      "89     UP000254640\n",
      "91     UP000001726\n",
      "93     UP000000556\n",
      "95     UP000000799\n",
      "98     UP000003917\n",
      "101    UP000005740\n",
      "113    UP000017248\n",
      "114    UP000001170\n",
      "120    UP000001978\n",
      "123    UP000001889\n",
      "125    UP000184310\n",
      "Name: Host, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Filter the DataFrame for bacteria\n",
    "bacteria_df = uniprot_reference[uniprot_reference[\"SUPERREGNUM\"] == \"bacteria\"]\n",
    "\n",
    "# Create a function for loose matching\n",
    "def find_proteome_id(name, species_column, proteome_column):\n",
    "    for index, species in enumerate(species_column):\n",
    "        if name in species:  # Check if the name is a substring of the species\n",
    "            return proteome_column.iloc[index]\n",
    "    return None\n",
    "\n",
    "# Map the bacteria names in the Series to their Proteome_ID using loose matching\n",
    "proteome_ids = bacteria_series.apply(\n",
    "    lambda name: find_proteome_id(name, bacteria_df[\"Species Name\"], bacteria_df[\"Proteome_ID\"])\n",
    ")\n",
    "\n",
    "# Drop NaN values and output the result\n",
    "proteome_ids = proteome_ids.dropna()\n",
    "print(len(proteome_ids))  # Count of matches\n",
    "print((proteome_ids))  # Matched Proteome_IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select the proteins listed in n_only_surface.gff3 and create n_only surface.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping 1: Missing files.\n",
      "Skipping 2: Missing files.\n",
      "Skipping 3: Missing files.\n",
      "Skipping 4: Missing files.\n",
      "Skipping 5: Missing files.\n",
      "Skipping 6: Missing files.\n",
      "Skipping 7: Missing files.\n",
      "Skipping 8: Missing files.\n",
      "Skipping 9: Missing files.\n",
      "Skipping 10: Missing files.\n",
      "Skipping 11: Missing files.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping 12: Missing files.\n",
      "Skipping 13: Missing files.\n",
      "Skipping 14: Missing files.\n",
      "Skipping 15: Missing files.\n",
      "Skipping 16: Missing files.\n",
      "Skipping 17: Missing files.\n",
      "Skipping 18: Missing files.\n",
      "Skipping 19: Missing files.\n",
      "Skipping 20: Missing files.\n",
      "Skipping 21: Missing files.\n",
      "Skipping 22: Missing files.\n",
      "Skipping 23: Missing files.\n",
      "Skipping 24: Missing files.\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[35], line 30\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[38;5;66;03m#print(headers)\u001b[39;00m\n\u001b[0;32m     26\u001b[0m \n\u001b[0;32m     27\u001b[0m \n\u001b[0;32m     28\u001b[0m     \u001b[38;5;66;03m# Read sequences from the FASTA file and filter by headers\u001b[39;00m\n\u001b[0;32m     29\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(fasta_file, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m fasta, \u001b[38;5;28mopen\u001b[39m(output_file, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m out_fasta:\n\u001b[1;32m---> 30\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m record \u001b[38;5;129;01min\u001b[39;00m \u001b[43mSeqIO\u001b[49m\u001b[38;5;241m.\u001b[39mparse(fasta, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfasta\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m     31\u001b[0m             \u001b[38;5;28;01mif\u001b[39;00m record\u001b[38;5;241m.\u001b[39mid \u001b[38;5;129;01min\u001b[39;00m headers:\n\u001b[0;32m     32\u001b[0m                 SeqIO\u001b[38;5;241m.\u001b[39mwrite(record, out_fasta, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfasta\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[1;32mIn[35], line 30\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[38;5;66;03m#print(headers)\u001b[39;00m\n\u001b[0;32m     26\u001b[0m \n\u001b[0;32m     27\u001b[0m \n\u001b[0;32m     28\u001b[0m     \u001b[38;5;66;03m# Read sequences from the FASTA file and filter by headers\u001b[39;00m\n\u001b[0;32m     29\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(fasta_file, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m fasta, \u001b[38;5;28mopen\u001b[39m(output_file, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m out_fasta:\n\u001b[1;32m---> 30\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m record \u001b[38;5;129;01min\u001b[39;00m \u001b[43mSeqIO\u001b[49m\u001b[38;5;241m.\u001b[39mparse(fasta, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfasta\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m     31\u001b[0m             \u001b[38;5;28;01mif\u001b[39;00m record\u001b[38;5;241m.\u001b[39mid \u001b[38;5;129;01min\u001b[39;00m headers:\n\u001b[0;32m     32\u001b[0m                 SeqIO\u001b[38;5;241m.\u001b[39mwrite(record, out_fasta, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfasta\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32m_pydevd_bundle/pydevd_cython.pyx:1457\u001b[0m, in \u001b[0;36m_pydevd_bundle.pydevd_cython.SafeCallWrapper.__call__\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32m_pydevd_bundle/pydevd_cython.pyx:701\u001b[0m, in \u001b[0;36m_pydevd_bundle.pydevd_cython.PyDBFrame.trace_dispatch\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32m_pydevd_bundle/pydevd_cython.pyx:1152\u001b[0m, in \u001b[0;36m_pydevd_bundle.pydevd_cython.PyDBFrame.trace_dispatch\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32m_pydevd_bundle/pydevd_cython.pyx:1135\u001b[0m, in \u001b[0;36m_pydevd_bundle.pydevd_cython.PyDBFrame.trace_dispatch\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32m_pydevd_bundle/pydevd_cython.pyx:312\u001b[0m, in \u001b[0;36m_pydevd_bundle.pydevd_cython.PyDBFrame.do_wait_suspend\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\lorenzo\\anaconda3\\envs\\esm\\lib\\site-packages\\debugpy\\_vendored\\pydevd\\pydevd.py:2070\u001b[0m, in \u001b[0;36mPyDB.do_wait_suspend\u001b[1;34m(self, thread, frame, event, arg, exception_type)\u001b[0m\n\u001b[0;32m   2067\u001b[0m             from_this_thread\u001b[38;5;241m.\u001b[39mappend(frame_custom_thread_id)\n\u001b[0;32m   2069\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_threads_suspended_single_notification\u001b[38;5;241m.\u001b[39mnotify_thread_suspended(thread_id, thread, stop_reason):\n\u001b[1;32m-> 2070\u001b[0m         keep_suspended \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_do_wait_suspend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mthread\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mframe\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mevent\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43marg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msuspend_type\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfrom_this_thread\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mframes_tracker\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   2072\u001b[0m frames_list \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m   2074\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m keep_suspended:\n\u001b[0;32m   2075\u001b[0m     \u001b[38;5;66;03m# This means that we should pause again after a set next statement.\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\lorenzo\\anaconda3\\envs\\esm\\lib\\site-packages\\debugpy\\_vendored\\pydevd\\pydevd.py:2106\u001b[0m, in \u001b[0;36mPyDB._do_wait_suspend\u001b[1;34m(self, thread, frame, event, arg, suspend_type, from_this_thread, frames_tracker)\u001b[0m\n\u001b[0;32m   2103\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_call_input_hook()\n\u001b[0;32m   2105\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprocess_internal_commands()\n\u001b[1;32m-> 2106\u001b[0m     \u001b[43mtime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msleep\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0.01\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m   2108\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcancel_async_evaluation(get_current_thread_id(thread), \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28mid\u001b[39m(frame)))\n\u001b[0;32m   2110\u001b[0m \u001b[38;5;66;03m# process any stepping instructions\u001b[39;00m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Folder containing the files\n",
    "folder_path = \"C:/Users/lorenzo/Desktop/Nuova cartella\"\n",
    "\n",
    "# Iterate through numbers corresponding to the files\n",
    "for n in range(1, 1902):\n",
    "    fasta_file = os.path.join(folder_path, f\"{n}.fasta\")\n",
    "    gff3_file = os.path.join(folder_path, f\"{n}_only_surface.gff3\")\n",
    "    output_file = os.path.join(folder_path, f\"{n}_only_surface.fasta\")\n",
    "\n",
    "    # Skip if files don't exist\n",
    "    if not os.path.exists(fasta_file) or not os.path.exists(gff3_file):\n",
    "        print(f\"Skipping {n}: Missing files.\")\n",
    "    \n",
    "\n",
    "    # Extract headers from the gff3 file\n",
    "    with open(gff3_file, 'r') as gff3:\n",
    "        headers = [\n",
    "            line.split(\" \")[1]\n",
    "            for line in gff3\n",
    "            if line.strip() and line.startswith(\"#\")\n",
    "        ]\n",
    "#print(headers)\n",
    "\n",
    "\n",
    "    # Read sequences from the FASTA file and filter by headers\n",
    "    with open(fasta_file, 'r') as fasta, open(output_file, 'w') as out_fasta:\n",
    "        for record in SeqIO.parse(fasta, \"fasta\"):\n",
    "            if record.id in headers:\n",
    "                SeqIO.write(record, out_fasta, \"fasta\")\n",
    "\n",
    "    print(f\"Created {output_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    # Read sequences from the FASTA file and filter by headers\n",
    "    with open(fasta_file, 'r') as fasta, open(output_file, 'w') as out_fasta:\n",
    "        for record in SeqIO.parse(fasta, \"fasta\"):\n",
    "            if record.id in headers:\n",
    "                SeqIO.write(record, out_fasta, \"fasta\")\n",
    "\n",
    "    print(f\"Created {output_file}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "esm",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
